File: 2019/Python/day8.py
### =============================================
### [ ADVENT OF CODE ] (https://adventofcode.com)
### 2019 - Mina Pêcheux: Python version
### ---------------------------------------------
### Day 8: Space Image Format
### =============================================

# [ Computation functions ]
# -------------------------
### PART I
def decode_layers(inputs, width, height):
    '''Decodes the image layers by splitting it in even chunks.
    
    :param inputs: Content of the image (as a string).
    :type inputs: str
    :param width: Width of each layer in the image.
    :type width: int
    :param height: Height of each layer in the image.
    :type height: int
    :return: Layers in the image.
    :rtype: list(str)
    '''
    layer_size = width * height
    n_layers = len(inputs) // layer_size
    return [ inputs[l*layer_size:(l+1)*layer_size] for l in range(n_layers) ]
    
def compute_checksum(inputs, width, height):
    '''Computes a basic checksum to verify the image is intact by finding the
    layer that has the fewest layer and computing the product of its number of
    1s and 2s.
    
    :param inputs: Content of the image (as a string).
    :type inputs: str
    :param width: Width of each layer in the image.
    :type width: int
    :param height: Height of each layer in the image.
    :type height: int
    :return: Checksum to verify the image validity.
    :rtype: int
    '''
    layers = decode_layers(inputs, width, height)
    best_n_zeros, best_value = None, None
    for layer in layers:
        n_zeros = layer.count('0')
        if best_n_zeros is None or n_zeros < best_n_zeros:
            n_ones = layer.count('1')
            n_twos = layer.count('2')
            best_n_zeros = n_zeros
            best_value = n_ones * n_twos
    return best_value, layers
    
### Part II
def display_message(layers, width, height):
    '''Displays the message that was sent (and has previously been divided into
    even layers).
    
    :param layers: Layers of the image.
    :type layers: list(str)
    :param width: Width of each layer in the image.
    :type width: int
    :param height: Height of each layer in the image.
    :type height: int
    '''
    marker = '█'
    # iterate through layers in reverse order to have the right depht overwrite
    img = [ [ ' ' for _ in range(width) ] for _ in range(height) ]
    for layer in layers[::-1]:
        # go through grid
        for y in range(height):
            for x in range(width):
                # and turn on/off pixels depending on the value (transparent has
                # no impact)
                if layer[x + y*width] == '1':
                    img[y][x] = marker
                elif layer[x + y*width] == '0':
                    img[y][x] = ' '
    # display the message
    print('')
    for y in range(height):
        print(''.join(img[y]))
    print('')
    
# [ Base tests ]
# --------------
def make_tests():
    '''Performs tests on the provided examples to check the result of the
    computation functions is ok.'''
    assert decode_layers('123456789012', 3, 2) == [ '123456', '789012' ]
    assert decode_layers('210012011212', 3, 2) == [ '210012', '011212' ]

    ### PART I
    c, _ = compute_checksum('123456789012', 3, 2)
    assert c == 1
    c, _ = compute_checksum('210012011212', 3, 2)
    assert c == 6

if __name__ == '__main__':
    # check function results on example cases
    make_tests()
    
    # get input data
    data_path = '../data/day8.txt'
    inputs = open(data_path, 'r').read().strip()
    
    ### PART I
    solution, layers = compute_checksum(inputs, 25, 6)
    print('PART I: solution = {}'.format(solution))
    
    ### PART II
    display_message(layers, 25, 6)
    print('PART II (see the shell)')
    

File: 2019/C++/src/day8.cpp
/**
 * \file day8.cpp
 * \brief AoC 2019 - Day 8 (C++ version)
 * \author Mina Pêcheux
 * \date 2019
 *
 * [ ADVENT OF CODE ] (https://adventofcode.com)
 * ---------------------------------------------
 * Day 8: Space Image Format
 * =============================================
 */
#include <algorithm>
#include "utils.hpp"

// [ Computation functions ]
// -------------------------

/**
 * \fn std::vector<std::string> decodeLayers(std::string inputs, int width, int height)
 * \brief Decodes the image layers by splitting it in even chunks.
 *
 * \param inputs Content of the image (as a string).
 * \param width Width of each layer in the image.
 * \param height Height of each layer in the image.
 * \return Layers in the image.
 */
std::vector<std::string> decodeLayers(std::string inputs, int width, int height) {
  std::vector<std::string> layers;
  int layerSize = width * height;
  int nLayers = inputs.length() / layerSize;
  for (int n = 0; n < nLayers; n++) {
    layers.push_back(inputs.substr(n*layerSize, layerSize));
  }
  return layers;
}

/*------------------------------------------------------------------------------
  Part I
------------------------------------------------------------------------------*/
/**
 * \fn int computeChecksum(std::vector<std::string>& layers, std::string inputs, int width, int height)
 * \brief Computes a basic checksum to verify the image is intact by finding the
 * layer that has the fewest layer and computing the product of its number of
 * 1s and 2s.
 *
 * \param layers Reference to the vector of layers (to fill).
 * \param inputs Content of the image (as a string).
 * \param width Width of each layer in the image.
 * \param height Height of each layer in the image.
 * \return Checksum to verify the image validity.
 */
int computeChecksum(std::vector<std::string>& layers, std::string inputs, int width, int height) {
  layers = decodeLayers(inputs, width, height);
  int bestNZeros = -1, bestValue = -1;
  int nZeros, nOnes, nTwos;
  for (auto layer : layers) {
    nZeros = std::count(layer.begin(), layer.end(), '0');
    if (bestNZeros == -1 || nZeros < bestNZeros) {
      nOnes = std::count(layer.begin(), layer.end(), '1');
      nTwos = std::count(layer.begin(), layer.end(), '2');
      bestNZeros = nZeros;
      bestValue = nOnes * nTwos;
    }
  }
  return bestValue;
}

/*------------------------------------------------------------------------------
  Part II
------------------------------------------------------------------------------*/
/**
 * \fn void displayMessage(std::vector<std::string>& layers, int width, int height)
 * \brief Displays the message that was sent (and has previously been divided into
   even layers).
 *
 * \param layers Layers of the image.
 * \param width Width of each layer in the image.
 * \param height Height of each layer in the image.
 */
void displayMessage(std::vector<std::string>& layers, int width, int height) {
  std::string marker = "█";
  // iterate through layers in reverse order to have the right depht overwrite
  std::string img[height][width];
  for (int l = layers.size() - 1; l >= 0; l--) {
    // go through grid
    for (int y = 0; y < height; y++) {
      for (int x = 0; x < width; x++) {
        // and turn on/off pixels depending on the value (transparent has no
        // impact)
        if (layers[l][x + y*width] == '1') {
          img[y][x] = marker;
        } else if (layers[l][x + y*width] == '0') {
          img[y][x] = " ";
        }
      }
    }
  }
  // display the message
  std::cout << "\n";
  for (int y = 0; y < height; y++) {
    for (int x = 0; x < width; x++) {
      std::cout << img[y][x];
    }
    std::cout << "\n";
  }
  std::cout << "\n";
}

// [ Base tests ]
// --------------

/**
 * \fn void makeTests()
 * \brief Performs tests on the provided examples to check the result of the
 * computation functions is ok.
 */
void makeTests() {
  std::vector<std::string> layers1;
  assert(computeChecksum(layers1, "123456789012", 3, 2) == 1);
  std::vector<std::string> layers2;
  assert(computeChecksum(layers2, "210012011212", 3, 2) == 6);
}

int main(int argc, char const *argv[]) {
  // check function results on example cases
  makeTests();

  // get input data
  std::string dataPath = "../data/day8.txt";
  std::string data = readFile(dataPath);
  std::vector<std::string> layers;
  
  // Part I
  int solution1 = computeChecksum(layers, data, 25, 6);
  std::cout << "PART I: solution = " << solution1 << '\n';
  
  // Part II
  displayMessage(layers, 25, 6);
  std::cout << "PART II (see the shell)" << '\n';
  
  return 0;
}
